Improve performance with cffi

MaxGyver83 opened the pull request:
Out of curiosity, I have analyzed the performance of *8vim_keyboard_layout_calculator* and tested how much faster this script can be when the bottleneck functions are rewritten in C and called via `python-cffi`.

This way the optimizer runs more than double as fast. See my article [Speed up Python with CFFI](https://maximilian-schillinger.de/articles/speed-up-python-with-cffi.html) for details.

I don't know if it makes sense merging this. Updating the C code when the Python code changes is additional maintenance work. No problem for me if you decline this pull request. At least, interested users might find this PR and apply the changes locally.

MaxGyver83 committed:
Call testSingleLayout in getLayoutScores

MaxGyver83 committed:
Rewrite testSingleLayout in C

MaxGyver83 committed:
Rewrite getTopScores in C

Glitchy-Tozier commented:
Thank you for the PR!
I haven't been able to try it out and test it yet, but I'll still address a few general points.

> I don't know if it makes sense merging this. Updating the C code when the Python code changes is additional maintenance work.

As I'm no longer trying to actively extend this optimizer, I don't mind some additional maintenance work (which won't affect me too much).
I think I'm fine with merging C-code if it's well documented (I've never coded in C before) and I first create a release to have a pre-C timestamp.

1. Just out of curiosity: Do you know whether the same thing would be possible using Rust? That language would be way more maintainable to me.
2. Over the course of the day I'll leave some comments on the code on things I'm wondering about when first reading the code. Feel free to push back on any suggestions you disagree with.

Glitchy-Tozier commented on main.py:
```diff
@@ -750,19 +766,26 @@ def getAsciiArray() -> array:
 def getLayoutScores(layouts: tuple, bigrams: tuple, prevScores=None) -> tuple:
     """Tests the layouts and return their scores. It's only used when single-threading."""
 
-    asciiArray = getAsciiArray()
     nrLayouts = len(layouts)
     scores = array("d", [0.0]*nrLayouts) # Create the empty scoring-list
+    if USE_CFFI:
+        bigr_count = len(bigrams)
+        bigr = ffi.new("Bigram[]", bigr_count)
+        for i, b in enumerate(bigrams):
+            bigr[i].letter1AsciiCode = b.letter1AsciiCode
+            bigr[i].letter2AsciiCode = b.letter2AsciiCode
+            bigr[i].frequency = b.frequency
+
+        # Test the flow of all the layouts.
+        for k, layout in enumerate(layouts):
+            char_list = layout.encode('latin1')
+            scores[k] = test_single_layout(char_list, len(layout), bigr, bigr_count, LINEAR_SCORE_LIST)
+    else:
+        asciiArray = getAsciiArray()
 
-    # Test the flow of all the layouts.
-    for k, layout in enumerate(layouts):
-        for j, letter in enumerate(layout):
-            asciiArray[ord(letter)] = j  # Fill up asciiArray
-
-        for bigram in bigrams: # Go through every bigram and see how well it flows.
-            firstLetterPlacement = asciiArray[bigram.letter1AsciiCode]
-            secondLetterPlacement = asciiArray[bigram.letter2AsciiCode]
-            scores[k] += bigram.frequency * SCORE_LIST[firstLetterPlacement][secondLetterPlacement]
+        # Test the flow of all the layouts.
+        for k, layout in enumerate(layouts):
+            scores[k] = testSingleLayout(layout, asciiArray, bigrams)
```
Doesn't this change negatively impact performance in non-cffi-cases? It adds thousands (millions?) of function calls.

MaxGyver83 replied:
Millions! If you have time, check my article linked in the PR description: I have tested it (PyPy, no CFFI) and it was surprisingly even a few seconds faster with these additional function calls. Maybe the PyPy JIT inlines these function calls and the little speed improvement was only variance!?

It's possible that this code would run slower in CPython. But since you recommend `USE_MULTIPROCESSING = True` for CPython, this function isn't called anyway.

Glitchy-Tozier replied:
Valid points, resolving this thread! :)

Glitchy-Tozier commented on main.py:
```diff
@@ -10,12 +11,27 @@
+              "Compile it with:\n"
+              f"cd cffi && {interpreter} ./cffi_extension_build.py")
+        sys.exit(1)
+
+    LINEAR_SCORE_LIST = list(itertools.chain.from_iterable(SCORE_LIST))
+
```
Why do we need to flatten the SCORE_LIST (I assume that's what's happening here)? Aren't nested arrays a thing in C?

MaxGyver83 replied:
This was just for simplicity. Flattening the list is a one-liner in Python and is executed only once. Passing a list of strings was more difficult than I thought. I don't know if it's easier for an array of arrays. I can try again, if you like.

Glitchy-Tozier replied:
It'd be neat if you could try. If you find that it complicates the code way too much, then just tell me and we'll stick to the flattened array.

MaxGyver83 replied:
I have replaced the `LINEAR_SCORE_LIST` with an ffi `"double[][32]"` object . See 1814d6cf4fb6a74c197c8c22be4eed3577713938. The "32" for the sub-array length is still hard-coded. (How likely will it change in the future?) Surprisingly, the optimizer runs now even faster. My last test run was 4m20s.

Glitchy-Tozier replied:
1. Looks neat!
2. The `32` is derived from the 32 positions on the keyboard. `SCORE_LIST` defines for each position how good each other position feels, thus 32×32.
As long as the keyboard keeps using 4 sections (=8 subsections) and 4 letters per subsection, `SCORE_LIST` will keep being a 32×32 matrix.
3. I imagine the speed improvement resulted from the now avoided multiplication/addition 🤔

MaxGyver83 replied:
> I imagine the speed improvement resulted from the now avoided multiplication/addition

Good point! This might be the reason.

Glitchy-Tozier commented on main.py:
```diff
@@ -10,12 +11,27 @@
+        from cffi._cffi_extension.lib import test_single_layout, get_top_scores
+        from cffi._cffi_extension import ffi
+    except ModuleNotFoundError as e:
+        implementation = platform.python_implementation()
+        if implementation == "CPython":
+            interpreter = "python3" if implementation == "CPython" else "pypy3"
```
Is there a reason for checking `if implementation == "CPython"` twice immediately after one another?

MaxGyver83 replied:
Oh, this is a bug! I forgot to delete line 26. I'll fix it.

MaxGyver83 replied:
Fixed in 5a3067162d021195008d5e5a38893692553219aa.

Glitchy-Tozier commented on main.py:
```diff
@@ -10,12 +11,27 @@
+if USE_CFFI:
+    try:
+        from cffi._cffi_extension.lib import test_single_layout, get_top_scores
+        from cffi._cffi_extension import ffi
```
This is a bit pedantic but why not name it "ffi" instead of "**c**ffi"?

MaxGyver83 replied:
Which of these `cffi`s? :-) Do you mean like this?

```py
if USE_FFI:
    try:
        from ffi._ffi_extension.lib import ...
```

Good question! I was even wondering if it shouldn't be "C extension" instead of "CFFI extension".

Glitchy-Tozier replied:
I just noticed I've written it the other way around... I meant to write:
`...why not name it "cffi" instead of "ffi"?`

I was referring to the name you assigned to the imported thing (it should be the last line I highlighted).

MaxGyver83 replied:
OK, now I understand. I didn't choose this name. It's always `ffi`. I saw this in the documentation, for example here:

> Then, in your main program, you use:
> ```py
> from _example import ffi, lib
> ```

<https://cffi.readthedocs.io/en/latest/overview.html>

Glitchy-Tozier replied:
Ok, then let's stick to their recommended naming scheme :)

Glitchy-Tozier commented on main.py:
```diff
@@ -750,19 +766,26 @@ def getAsciiArray() -> array:
     scores = array("d", [0.0]*nrLayouts) # Create the empty scoring-list
+    if USE_CFFI:
+        bigr_count = len(bigrams)
+        bigr = ffi.new("Bigram[]", bigr_count)
```
Am I correct in assuming this is an array of the C-struct-`Bigram`s that's getting created?

Could you...
1. ... rename the struct to make it obvious that this is not the Python-`Bigram` being used?
2. ... in general decide on a naming scheme for all variables that immediately highlights whether the thing
    1. Is a python variable or function that is only required due to cffi (for example `bigr`)
    2. Is a c variable or function that now gets called from the python code (for example `test_single_layout()`)

MaxGyver83 replied:
> Am I correct in assuming this is an array of the C-struct-`Bigram`s that's getting created?

Yes.

> Could you...
> 
>     1. ... rename the struct to make it obvious that this is not the Python-`Bigram` being used?

Sure.

>     2. ... in general decide on a naming scheme for all variables that immediately highlights whether the thing
>        
>        1. Is a python variable or function that is only required due to cffi (for example `bigr`)
>        2. Is a c variable or function that now gets called from the python code (for example `test_single_layout()`)

Yes, sounds reasonable.

MaxGyver83 replied:
I have renamed the ffi variables. See commit ec64525a15d1c4b98402e12abddfbe8b7f4653c2.

Glitchy-Tozier replied:
The naming looks really good to me :)

One further thing: Could you check whether the change in calling `test_single_layout` impacts performance? I've heard of the notion that doing 
```py
x = a.b.c
for range(10000):
    x()
```
is faster (in non-compiled languages such as Python) than
```py
for range(10000):
    a.b.c()
```
Got any insights regarding that idea?

MaxGyver83 replied:
I have tested with this commit (ec64525a15d1c4b98402e12abddfbe8b7f4653c2). It took 6m10s. A bit slower than before. But I'm not really sure how reproducible it is. I haven't tested in a clean environment (p.e. Firefox is running, too). Maybe it's true that Python now "searches" for the right function in every loop iteration.

MaxGyver83 replied:
Another run: I have undone the change regarding the C function imports, based on the now latest commit (1814d6cf4fb6a74c197c8c22be4eed3577713938). And it took again 4m20s. (Like [this test](https://github.com/Glitchy-Tozier/8vim_keyboard_layout_calculator/pull/12#discussion_r1196958021).)

Glitchy-Tozier replied:
Are you saying your tests validate the notion I mentioned above (that `x()` is faster than `a.b.c()`)?

MaxGyver83 replied:
At first, I thought yes. But in my second test, it didn't make any difference. In the second test, I compared commit 1814d6c with these local changes:

```diff
diff --git a/main.py b/main.py
index 99dfc4c..dd1e145 100644
--- a/main.py
+++ b/main.py
diff --git a/main.py b/main.py
index 99dfc4c..dd1e145 100644
--- a/main.py
+++ b/main.py
@@ -19,7 +19,7 @@ start_time = time()

 if USE_CFFI:
     try:
-        from cffi import _cffi_extension
+        from cffi._cffi_extension.lib import test_single_layout, get_top_scores
         from cffi._cffi_extension import ffi
     except ModuleNotFoundError as e:
         implementation = platform.python_implementation()
@@ -779,7 +779,7 @@ def getLayoutScores(layouts: tuple, bigrams: tuple, prevScores=None) -> tuple:
         # Test the flow of all the layouts.
         for k, layout in enumerate(layouts):
             permutatedLayoutBytes = layout.encode('latin1')
-            scores[k] = _cffi_extension.lib.test_single_layout(
+            scores[k] = test_single_layout(
                 permutatedLayoutBytes, len(layout), ffiBigrams, nrBigrams, FFI_SCORE_LIST
             )
     else:
@@ -806,7 +806,7 @@ def getLayoutScores(layouts: tuple, bigrams: tuple, prevScores=None) -> tuple:
         ffiGoodLayouts = [ffi.new("char[]", 100)] * 500
         ffiGoodLayoutsPointer = ffi.new("char*[]", ffiGoodLayouts)
         ffiGoodScores = ffi.new('double[]', 500)
-        _cffi_extension.lib.get_top_scores(
+        get_top_scores(
             ffiLayoutsPointer, nrLayouts, ffiScoresArray, 500, LETTERS_PER_LAYER, ffiGoodLayoutsPointer, ffiGoodScores
         )
         return [ffi.string(s).decode("latin1") for s in ffiGoodLayoutsPointer], [float(f) for f in ffiGoodScores]
@@ -925,7 +925,7 @@ def greedyOptimization(layouts: tuple, scores: array, info: InfoWithTime = None)
                 layoutPermutations = performLetterSwaps(layout)
                 for i, permutatedLayout in enumerate(layoutPermutations):
                     permutatedLayoutBytes = permutatedLayout.encode('latin1')
-                    permutatedScore = _cffi_extension.lib.test_single_layout(
+                    permutatedScore = test_single_layout(
                         permutatedLayoutBytes, len(permutatedLayout), ffiBigrams, nrBigrams, FFI_SCORE_LIST
                     )
                     if permutatedScore > score:
```

Glitchy-Tozier replied:
I see, maybe pypy is able to cache the re-loading.

1. We can leave this part of the code as it currently is (`a.b.c()`)
2. As we've seen the results vary, could you go to `def greedyOptimization()` and double-check whether moving the `if` inside the for-loop causes a slowdown?

MaxGyver83 replied:
>     1. We can leave this part of the code as it currently is (`a.b.c()`)

OK, fine. I thought you wanted it that explicit to make clear we are calling C functions here. Maybe I misunderstood you. Of course, you can change it anytime, when you think it's inefficient this way.

>     2. As we've seen the results vary, could you go to `def greedyOptimization()` and double-check whether moving the `if` inside the for-loop causes a slowdown?

I have tried again. With the latest commit, it took 4m16s. With `if USE_CFFI` inside the while loop, now it took 4m18s.

Glitchy-Tozier replied:
> OK, fine. I thought you wanted it that explicit to make clear we are calling C functions here. Maybe I misunderstood you. Of course, you can change it anytime, when you think it's inefficient this way.

You got it right that I wanted to make it clear that we're calling C-functions here. It's just that I was thinking about something in the vicinity of this:
```py
import cffi.some_function as cSomeFunction
...
cSomeFunction()
```

> > 2. As we've seen the results vary, could you go to `def greedyOptimization()` and double-check whether moving the `if` inside the for-loop causes a slowdown?
> 
> I have tried again. With the latest commit, it took 4m16s. With `if USE_CFFI` inside the while loop, now it took 4m18s.

Thank you for re-testing! In that case, let's move the `if` inside the loop. ~2s seem very tame, and the code gets easier to read/maintain this way.

MaxGyver83 replied:
OK, see commit 9922bba.

Glitchy-Tozier commented on main.py:
```diff
@@ -867,25 +900,46 @@ def greedyOptimization(layouts: tuple, scores: array, info: InfoWithTime = None)
+            while optimizing is True:
+                layoutPermutations = performLetterSwaps(layout)
+                for i, permutatedLayout in enumerate(layoutPermutations):
+                    char_list = permutatedLayout.encode('latin1')
+                    permutatedScore = test_single_layout(char_list, len(permutatedLayout), bigr, bigr_count, LINEAR_SCORE_LIST)
```
There two lines seem to be the only difference between cffi and regular code, which leads to tons of code-duplication. Could we move the if-check right before these lines?

I'm fine with the optimizer taking one second longer if it avoids duplicated code.

MaxGyver83 replied:
Good point. I thought it might be inefficient to check the value of `USE_CFFI` over and over again within the `while` loop. But I didn't test it. So I guess this was a case of premature optimization.

MaxGyver83 replied:
I have tested it. Checking the value of `USE_CFFI` inside the loop increases the run time from 5m56s to 6m13 (that's ≈ +5%). Is it worth changing it?

Glitchy-Tozier replied:
Hmmmm that's unfortunate. I didn't think checking a boolean variable would add that much computation time.
Let's leave it the way you wrote it!

Glitchy-Tozier commented on cffi/cffi_extension_build.py:
```diff
@@ -0,0 +1,24 @@
+ffibuilder = FFI()
+
+ffibuilder.cdef("""\
+typedef struct Bigram {
+    int letter1AsciiCode;
+    int letter2AsciiCode;
+    double frequency;
+} Bigram;
+double test_single_layout(char* layout, int layout_length, Bigram* bigrams,
+                          int bigrams_count, double* score_list);
+void get_top_scores(char** layouts, long layout_count, double* scores, int nr_of_best,
+                    int letters_per_layer, char** top_layouts, double* top_scores);
+""")
```
This is the same code as in the otther file, correct? Couldn't we read the file instead of having the code present in two places?

MaxGyver83 replied:
I thought the same. I was wondering about the same type of duplication in the python-cffi [documentation](https://cffi.readthedocs.io/en/latest/overview.html#api-mode-calling-c-sources-instead-of-a-compiled-library). Maybe they do it like this in the example because it's only one line.
I'll change this.

MaxGyver83 replied:
Done in 3247aff9871ed271bd5412607df638b6d869c05f.

MaxGyver83 commented:
>     1. Just out of curiosity: Do you know whether the same thing would be possible using Rust? That language would be way more maintainable to me.

I think this is possible and it would look almost the same on the Python side. I have never tried but I found this example: [Calling Rust from Python/PyPy using CFFI](https://gist.github.com/seanjensengrey/f5d73bbdf22cfa1ad463)

>     2. Over the course of the day I'll leave some comments on the code on things I'm wondering about when first reading the code. Feel free to push back on any suggestions you disagree with.

Thanks for your time and your comments!

MaxGyver83 committed:
Bug fix: Delete outdated if condition

MaxGyver83 committed:
cffi_extension_build.py: Read C header instead duplicating it

Glitchy-Tozier commented:
> I think this is possible and it would look almost the same on the Python side. I have never tried but I found this example: Calling Rust from Python/PyPy using CFFI

In case you are able to change, it would be neat if we could use Rust.

If you don't feel experienced enough or figure it would take too long, let's leave it using C-code. Unfortunately, I'm too overwhelmed by university at the moment to delve into this issue myself.

Glitchy-Tozier commented:
By the way, it would feel even more fun to read your article if you mentioned the time it takes for the script to run using `python3`. "Very long → 13min" doesn't give (me) the same feeling of accomplishment as "XXmin → 13min".
![Screenshot_20230512-233607_Firefox Klar.jpg](https://github.com/Glitchy-Tozier/8vim_keyboard_layout_calculator/assets/59611881/43920ec2-ef06-4f60-88ee-592b7d594eba)

MaxGyver83 commented:
> By the way, it would feel even more fun to read your article if you mentioned the time it takes for the script to run using `python3`. "Very long → 13min" doesn't give (me) the same feeling of accomplishment as "XXmin → 13min". 

I'll add a duration for `python3` when I get one. I tried twice but canceled after some time. I wasn't sure if it was still running at all. Maybe my laptop ran out of memory!? I'll try again.

MaxGyver83 commented:
> > I think this is possible and it would look almost the same on the Python side. I have never tried but I found this example: Calling Rust from Python/PyPy using CFFI
> 
> In case you are able to change, it would be neat if we could use Rust.
> 
> If you don't feel experienced enough or figure it would take too long, let's leave it using C-code. Unfortunately, I'm too overwhelmed by university at the moment to delve into this issue myself.

My Rust knowledge is also very basic. I'm interested in this language, but it's not at top of the list of things I want to do when I have some free time. :-)

Glitchy-Tozier commented:
> My Rust knowledge is also very basic. I'm interested in this language, but it's not at top of the list of things I want to do when I have some free time. :-)

In that case, let's stick to the C-code for now.
Should you ever feel motivated enough, replacing the c code with rust code would be greatly appreciated.

MaxGyver83 committed:
Rename variables; don't import C functions directly

- Use mixedCase for all variables in Python (for consistency)
- Prepend all ffi variables with 'ffi'
- Rename the 'Bigram' struct in C to 'BigramC'

MaxGyver83 committed:
Replace LINEAR_SCORE_LIST with ffi 'double[][32]' object

Glitchy-Tozier commented:
By the way, could you do a few further things:

For me (=maintainability):
- [x] Please add comments to the C-code. It would be great to have the purpose of each file and function documented (even though your great naming helps quite a bit). Additional comments mid-code also would be great, as you see fit.
- [x] It is especially important to add comments in any intermediary programming steps that differ from the python code you're trying to recreate.
- [x] Please create a short README-file in the cffi-directory that explains what cffi is, what parts of the code were extracted and why it was done. Feel free to also refer to your article or any other (easily understandable) resources.

For users:
- [x] In the main README, please mention the additional steps necessary to get the optimizer running. Mention that the prompt only needs to be run once.
- [x] In the error message in main.py, please mention that there is the option to turn off cffi mode. (If you have objections to that point let me know, I'm not 100% convinced of it myself.)

MaxGyver83 committed:
Add CFFI compile step to README; default USE_CFFI to False; add info how to disable CFFI in import error message

MaxGyver83 committed:
Add function comments in cffi_extension.h

MaxGyver83 committed:
Add README file for CFFI

MaxGyver83 committed:
More comments in cffi_extension.c

MaxGyver83 committed:
Move check of USE_CFFI inside while loop

MaxGyver83 commented:
>     * [ ]  Please add comments to the C-code. It would be great to have the purpose of each file and function documented (even though your great naming helps quite a bit). Additional comments mid-code also would be great, as you see fit.
> 
>     * [ ]  It is especially important to add comments in any intermediary programming steps that differ from the python code you're trying to recreate.

C comments added in 5e79c9a.

Glitchy-Tozier commented:
I think I'm slowly running out of things to complain about ;)
I've now started with testing. Three things:

1. Quick question: What was the thought process behind using `USE_CFFI = False` by default? To avoid crashes for preexisting users?
I think this PR is good enough to leave this Option enabled by default.

---

2. I saw that it is also required to do `pip install cffi`. How about we also add that info to the main README, something like this command:
```sh
cd cffi && pip install cffi && pypy3 cffi_extension_build.py ; cd ..
```
(The last point is to return to the root, allowing for the next command to be run seamlessly. `;` to do so even if previous steps fail.)

---

3. On that note, I receive this error when trying to do `pypy3 cffi_extension_build.py` (I'm testing on Linux, if that makes any difference):
```sh
$ pypy3 cffi_extension_build.py
generating ./_cffi_extension.c
(already up-to-date)
the current directory is '/home/florian/Downloads/8vim_keyboard_layout_calculator-improve-performance-with-cffi/cffi'
running build_ext
building '_cffi_extension' extension
gcc -pthread -DNDEBUG -O2 -fPIC -I/usr/include/pypy3.8 -c _cffi_extension.c -o ./_cffi_extension.o
_cffi_extension.c:50:14: fatal error: pyconfig.h: No such file or directory
   50 | #    include <pyconfig.h>
      |              ^~~~~~~~~~~~
compilation terminated.
Traceback (most recent call last):
  File "/usr/lib/pypy3.8/distutils/unixccompiler.py", line 133, in _compile
    extra_postargs)
  File "/usr/lib/pypy3.8/distutils/ccompiler.py", line 910, in spawn
    spawn(cmd, dry_run=self.dry_run)
  File "/usr/lib/pypy3.8/distutils/spawn.py", line 36, in spawn
    _spawn_posix(cmd, search_path, dry_run=dry_run)
  File "/usr/lib/pypy3.8/distutils/spawn.py", line 159, in _spawn_posix
    % (cmd, exit_status))
distutils.errors.DistutilsExecError: command 'gcc' failed with exit status 1

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/pypy3.8/cffi/ffiplatform.py", line 51, in _build
    dist.run_command('build_ext')
  File "/usr/lib/pypy3.8/distutils/dist.py", line 985, in run_command
    cmd_obj.run()
  File "/usr/lib/pypy3.8/distutils/command/build_ext.py", line 343, in run
    self.build_extensions()
  File "/usr/lib/pypy3.8/distutils/command/build_ext.py", line 452, in build_extensions
    self._build_extensions_serial()
  File "/usr/lib/pypy3.8/distutils/command/build_ext.py", line 477, in _build_extensions_serial
    self.build_extension(ext)
  File "/usr/lib/pypy3.8/distutils/command/build_ext.py", line 537, in build_extension
    depends=ext.depends)
  File "/usr/lib/pypy3.8/distutils/ccompiler.py", line 574, in compile
    self._compile(obj, src, ext, cc_args, extra_postargs, pp_opts)
  File "/usr/lib/pypy3.8/distutils/unixccompiler.py", line 135, in _compile
    raise CompileError(msg)
distutils.errors.CompileError: command 'gcc' failed with exit status 1

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "cffi_extension_build.py", line 17, in <module>
    ffibuilder.compile(verbose=True)
  File "/usr/lib/pypy3.8/cffi/api.py", line 727, in compile
    compiler_verbose=verbose, debug=debug, **kwds)
  File "/usr/lib/pypy3.8/cffi/recompiler.py", line 1565, in recompile
    compiler_verbose, debug)
  File "/usr/lib/pypy3.8/cffi/ffiplatform.py", line 22, in compile
    outputfilename = _build(tmpdir, ext, compiler_verbose, debug)
  File "/usr/lib/pypy3.8/cffi/ffiplatform.py", line 58, in _build
    raise VerificationError('%s: %s' % (e.__class__.__name__, e))
cffi.VerificationError: CompileError: command 'gcc' failed with exit status 1
```

Edit: `"This will use a C compiler in the background. If you don't have one installed, install one and try again."` Could this be the issue, I think I never explicitly installed a C-compiler. If this step is required, I think we should
- Add that info to the README
- Maybe keep the cffi-option off by default, since setup is more complicated than anticipated.

MaxGyver83 commented:
>     1. Quick question: What was the thought process behind using `USE_CFFI = False` by default? To avoid crashes for preexisting users?
>        I think this PR is good enough to leave this Option enabled by default.

My thought was that when `USE_CFFI` defaults to `False`, the calculator works out of the box. When it defaults to `True`, the user has to either compile the C extension or to set `USE_CFFI` to `True`.

>     * Maybe keep the cffi-option off by default, since setup is more complicated than anticipated.

OK.

>     2. I saw that it is also required to do `pip install cffi`. How about we also add that info to the main README, something like this command:
> 
> 
> ```shell
> cd cffi && pip install cffi && pypy3 cffi_extension_build.py ; cd ..
> ```
> 
> (The last point is to return to the root, allowing for the next command to be run seamlessly. `;` to do so even if previous steps fail.)

AFAIU, `cffi` is included in a PyPy installation. It only has to be installed, when somebody wants to use CFFI with CPython.
OK, I'll add a `cd ..`.

>     3. On that note, I receive this error when trying to do `pypy3 cffi_extension_build.py` (I'm testing on Linux, if that makes any difference):
> 
> 
> ```shell
> $ pypy3 cffi_extension_build.py
> generating ./_cffi_extension.c
> (already up-to-date)
> the current directory is '/home/florian/Downloads/8vim_keyboard_layout_calculator-improve-performance-with-cffi/cffi'
> running build_ext
> building '_cffi_extension' extension
> gcc -pthread -DNDEBUG -O2 -fPIC -I/usr/include/pypy3.8 -c _cffi_extension.c -o ./_cffi_extension.o
> _cffi_extension.c:50:14: fatal error: pyconfig.h: No such file or directory
>    50 | #    include <pyconfig.h>
>       |              ^~~~~~~~~~~~
> compilation terminated.
> Traceback (most recent call last):
>   File "/usr/lib/pypy3.8/distutils/unixccompiler.py", line 133, in _compile
>     extra_postargs)
>   File "/usr/lib/pypy3.8/distutils/ccompiler.py", line 910, in spawn
>     spawn(cmd, dry_run=self.dry_run)
>   File "/usr/lib/pypy3.8/distutils/spawn.py", line 36, in spawn
>     _spawn_posix(cmd, search_path, dry_run=dry_run)
>   File "/usr/lib/pypy3.8/distutils/spawn.py", line 159, in _spawn_posix
>     % (cmd, exit_status))
> distutils.errors.DistutilsExecError: command 'gcc' failed with exit status 1
> 
> During handling of the above exception, another exception occurred:
> 
> Traceback (most recent call last):
>   File "/usr/lib/pypy3.8/cffi/ffiplatform.py", line 51, in _build
>     dist.run_command('build_ext')
>   File "/usr/lib/pypy3.8/distutils/dist.py", line 985, in run_command
>     cmd_obj.run()
>   File "/usr/lib/pypy3.8/distutils/command/build_ext.py", line 343, in run
>     self.build_extensions()
>   File "/usr/lib/pypy3.8/distutils/command/build_ext.py", line 452, in build_extensions
>     self._build_extensions_serial()
>   File "/usr/lib/pypy3.8/distutils/command/build_ext.py", line 477, in _build_extensions_serial
>     self.build_extension(ext)
>   File "/usr/lib/pypy3.8/distutils/command/build_ext.py", line 537, in build_extension
>     depends=ext.depends)
>   File "/usr/lib/pypy3.8/distutils/ccompiler.py", line 574, in compile
>     self._compile(obj, src, ext, cc_args, extra_postargs, pp_opts)
>   File "/usr/lib/pypy3.8/distutils/unixccompiler.py", line 135, in _compile
>     raise CompileError(msg)
> distutils.errors.CompileError: command 'gcc' failed with exit status 1
> 
> During handling of the above exception, another exception occurred:
> 
> Traceback (most recent call last):
>   File "cffi_extension_build.py", line 17, in <module>
>     ffibuilder.compile(verbose=True)
>   File "/usr/lib/pypy3.8/cffi/api.py", line 727, in compile
>     compiler_verbose=verbose, debug=debug, **kwds)
>   File "/usr/lib/pypy3.8/cffi/recompiler.py", line 1565, in recompile
>     compiler_verbose, debug)
>   File "/usr/lib/pypy3.8/cffi/ffiplatform.py", line 22, in compile
>     outputfilename = _build(tmpdir, ext, compiler_verbose, debug)
>   File "/usr/lib/pypy3.8/cffi/ffiplatform.py", line 58, in _build
>     raise VerificationError('%s: %s' % (e.__class__.__name__, e))
> cffi.VerificationError: CompileError: command 'gcc' failed with exit status 1
> ```

This error message doesn't look like `gcc` is missing. In this case, I would expect something like `gcc: command not found`. But you can check if `gcc` is installed by running `gcc --version`.

Which distribution do you use? You could try installing the `python-dev` package if available.

MaxGyver83 committed:
cffi/README.md: Tell user to `cd ..` when done in cffi folder

Glitchy-Tozier commented:
> AFAIU, cffi is included in a PyPy installation. It only has to be installed, when somebody wants to use CFFI with CPython.

Interesting, did't know about that.


> This error message doesn't look like gcc is missing. In this case, I would expect something like gcc: command not found. But you can check if gcc is installed by running gcc --version.

You're right, it already is installed:

```sh
$ gcc --version
gcc (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0
Copyright (C) 2021 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

> Which distribution do you use?

Linux Mint (based on Ubuntu).

> You could try installing the python-dev package if available.

Which one do you mean and how would it benefit the process?
![image](https://github.com/Glitchy-Tozier/8vim_keyboard_layout_calculator/assets/59611881/b4d3ddcc-27ec-4267-9d0a-aa9383168072)

MaxGyver83 commented:
Maybe it's named `python3-dev`!?

Your error message looks like `pyconfig.h` is needed. On my laptop it's located here:

```
/opt/pypy3/include/pypy3.9/pyconfig.h
```

and here

```
/usr/include/python3.11/pyconfig.h
```

Maybe it's already installed but `gcc` can't find it. You could try searching for it:

```sh
fd pyconfig.h /
```

If you don't have [fd](https://github.com/sharkdp/fd) installed, then try:

```sh
find / -name pyconfig.h 2>/dev/null
```

Glitchy-Tozier commented:
> Maybe it's named `python3-dev`!?

Yep, that one exists and I already had it installed.


> Your error message looks like `pyconfig.h` is needed. On my laptop it's located here:
> 
> ```
> /opt/pypy3/include/pypy3.9/pyconfig.h
> ```
> 
> and here
> 
> ```
> /usr/include/python3.11/pyconfig.h
> ```
> 
> Maybe it's already installed but `gcc` can't find it. You could try searching for it:
> 
> ```shell
> fd pyconfig.h /
> ```
> 
> If you don't have [fd](https://github.com/sharkdp/fd) installed, then try:
> 
> ```shell
> find / -name pyconfig.h 2>/dev/null
> ```

I got this. What does it tell us?
```sh
$ find / -name pyconfig.h 2>/dev/null
/var/lib/flatpak/runtime/org.freedesktop.Sdk/x86_64/22.08/.74a00ed8035dcdea3f18a098764adfe0243a7bb5e6532caee6c8c14b2ce010a9-1H2BZ1/files/include/python3.10/pyconfig.h
/var/lib/flatpak/runtime/org.freedesktop.Sdk/x86_64/22.08/.74a00ed8035dcdea3f18a098764adfe0243a7bb5e6532caee6c8c14b2ce010a9-1H2BZ1/files/lib/x86_64-linux-gnu/include/python3.10/pyconfig.h
/var/lib/flatpak/runtime/org.freedesktop.Sdk/x86_64/22.08/2652112cf6f02f590602d3379a36a0e59d801a120f8ba4bb43170e75e3ab73da/files/include/python3.10/pyconfig.h
/var/lib/flatpak/runtime/org.freedesktop.Sdk/x86_64/22.08/2652112cf6f02f590602d3379a36a0e59d801a120f8ba4bb43170e75e3ab73da/files/lib/x86_64-linux-gnu/include/python3.10/pyconfig.h
/var/lib/flatpak/runtime/org.freedesktop.Sdk/x86_64/20.08/52034b76d4c302e58dd0c5b2b46d00b9f33a68cbe56859a264e426880e00979c/files/include/python3.8/pyconfig.h
/var/lib/flatpak/runtime/org.freedesktop.Sdk/x86_64/20.08/52034b76d4c302e58dd0c5b2b46d00b9f33a68cbe56859a264e426880e00979c/files/lib/x86_64-linux-gnu/include/python3.8/pyconfig.h
/var/lib/flatpak/runtime/org.freedesktop.Sdk/x86_64/21.08/.a439c06472701e00aead26fc20f474b187604a7803d942c56908743ef54e9c61-LN80X1/files/include/python3.9/pyconfig.h
/var/lib/flatpak/runtime/org.freedesktop.Sdk/x86_64/21.08/.a439c06472701e00aead26fc20f474b187604a7803d942c56908743ef54e9c61-LN80X1/files/lib/x86_64-linux-gnu/include/python3.9/pyconfig.h
/var/lib/flatpak/runtime/org.freedesktop.Sdk/x86_64/21.08/00ad08145692e66d6d7eff58c56299ac1e53b383877ba2c12497bc21cf16f946/files/include/python3.9/pyconfig.h
/var/lib/flatpak/runtime/org.freedesktop.Sdk/x86_64/21.08/00ad08145692e66d6d7eff58c56299ac1e53b383877ba2c12497bc21cf16f946/files/lib/x86_64-linux-gnu/include/python3.9/pyconfig.h

(...Many copies in timeshift snapshots...)

/usr/include/x86_64-linux-gnu/python3.10/pyconfig.h
/usr/include/python3.10/pyconfig.h
```

Glitchy-Tozier commented:
By the way, I just realized: Using `python3`, it _does_ compile:
```sh
$ cd cffi && pip install cffi && python3 cffi_extension_build.py ; cd ..

Defaulting to user installation because normal site-packages is not writeable
Collecting cffi
  Using cached cffi-1.15.1-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (441 kB)
Requirement already satisfied: pycparser in /home/florian/.local/lib/python3.10/site-packages (from cffi) (2.21)
Installing collected packages: cffi
Successfully installed cffi-1.15.1
generating ./_cffi_extension.c
(already up-to-date)
the current directory is '/home/florian/Downloads/8vim_keyboard_layout_calculator-improve-performance-with-cffi/cffi'
running build_ext
building '_cffi_extension' extension
x86_64-linux-gnu-gcc -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -g -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -I/usr/include/python3.10 -c _cffi_extension.c -o ./_cffi_extension.o
x86_64-linux-gnu-gcc -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -g -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -I/usr/include/python3.10 -c cffi_extension.c -o ./cffi_extension.o
x86_64-linux-gnu-gcc -shared -Wl,-O1 -Wl,-Bsymbolic-functions -Wl,-Bsymbolic-functions -g -fwrapv -O2 -Wl,-Bsymbolic-functions -g -fwrapv -O2 -g -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 ./_cffi_extension.o ./cffi_extension.o -o ./_cffi_extension.cpython-310-x86_64-linux-gnu.so
```
It can then be started via `python3 main.py`.

The previous error only occurs when using `pypy3` for the build script.

MaxGyver83 commented:
OK, no matches in a `pypy` folder. I suppose you'll have to install `pypy3-dev` to get the header files for PyPy, too.

Glitchy-Tozier commented:
Hey, that did the trick! :)

In that case, to (probably) conclude this PR, could you...
1. Move the instructions from the main README into the cffi-README and link to it (while keeping the mention of much faster execution time)
2. Extend the instructions in the cffi-README to mention those `{interpreter}-dev`-packages. I've written up a suggestion, a markdown-table that seems to be pretty compact but mentions most of the necessary information ([Here's the generator.](https://www.tablesgenerator.com/markdown_tables)):

```
| 1. Decide on Interpreter         | pypy3<br>(recommended)                                 | python3                                                  |
|:---------------------------------|--------------------------------------------------------|----------------------------------------------------------|
| 2. Install package (Linux only?) | pypy3-dev                                              | python3-dev                                              |
| 3. Get `cffi`                    | Included by default                                    | Run `pip install cffi`                                   |
| 4. Build code                    | Run `cd cffi && pypy3 cffi_extension_build.py ; cd ..` | Run `cd cffi && python3 cffi_extension_build.py ; cd ..` |
| 5. Start optimizer               | Run `pypy3 main.py`                                    | Run `python3 main.py`                                    |
```

| 1. Decide on Interpreter         | pypy3<br>(recommended)                                 | python3                                                  |
|:---------------------------------|--------------------------------------------------------|----------------------------------------------------------|
| 2. Install package (Linux only?) | pypy3-dev                                              | python3-dev                                              |
| 3. Get `cffi`                    | Included by default                                    | Run `pip install cffi`                                   |
| 4. Build code                    | Run `cd cffi && pypy3 cffi_extension_build.py ; cd ..` | Run `cd cffi && python3 cffi_extension_build.py ; cd ..` |
| 5. Start optimizer               | Run `pypy3 main.py`                                    | Run `python3 main.py`                                    |

MaxGyver83 commented:
Great overview! I'll update the READMEs when I'm back home.

Glitchy-Tozier commented:
Great! Feel free to edit it if you feel like making any changes / have additions.
(When editing, keep that colon (`:-------------------`) intact to preserve left-alignment.)

Glitchy-Tozier commented:
Neat! one typo-fix and we're done! :)
![Screenshot_20230521-205636_GitHub.jpg](https://github.com/Glitchy-Tozier/8vim_keyboard_layout_calculator/assets/59611881/f8812a61-d74d-4c8d-8508-1be04837eb20)

MaxGyver83 committed:
Improve cffi/README.md

MaxGyver83 commented:
> Neat! one typo-fix and we're done! :)

Fixed!

Glitchy-Tozier commented:
Well then let's merge!
Thank you very much for the hard work and the patience when working through this PR. :)
